File: haversine.cpp
#include <stdio.h> /* printf, fgets */
#include <stdlib.h> /* atof */
#include <math.h> /* sin */
#include <stdint.h>

float toRad(float x) {
  return x * 3.141592653 / 180;
}

float haversine(float lat1, float lon1, float lat2, float lon2, bool metric) {
  float R;
  if (metric) R = 6371; // km
  else R = 3958.756; // mi
  float x1 = lat2 - lat1;
  float dLat = toRad(x1);
  float x2 = lon2 - lon1;
  float dLon = toRad(x2);
  float a =
    sin(dLat / 2) * sin(dLat / 2) +
    cos(toRad(lat1)) * cos(toRad(lat2)) *
    sin(dLon / 2) * sin(dLon / 2);
  float c = 2 * atan2(sqrt(a), sqrt(1 - a));
  float d = R * c;
  return round(d * 100.0) / 100;
}

int main(int argc, char **argv) {
  float lat0, lat1, long0, long1;
  if (argc<5) {
    printf("Default coords. argc = %d\n", argc);
    lat0 = 22.4;
    lat1 = 22.4;
    long0 = 113.9;
    long1 = 114.0;
  } else {
    printf("Custom coords. argc = %d\n", argc);
    lat0 = atof(argv[1]);
    long0 = atof(argv[2]);
    lat1 = atof(argv[3]);
    long1 = atof(argv[4]);
  }
  float km = haversine(lat0, long0, lat1, long1, true);
  float mi = haversine(lat0, long0, lat1, long1, false);
  printf("distance between %.6f, %.6f and %.6f, %.6f:\n * %.3f km\n * %.3f mi\n", lat0, long0, lat1, long1, km, mi);
  return 0;
}

File: haversine.py
import sys, math

def toRad(x):
  return x * 3.141592653 / 180

def haversine(lat1, lon1, lat2, lon2, metric = True):
  R = 6371.0
  if (metric != True):
    R = 3958.756
  x1 = lat2-lat1
  dLat = toRad(x1)
  x2 = lon2-lon1
  dLon = toRad(x2)
  a = math.sin(dLat/2) * math.sin(dLat/2) + math.cos(toRad(lat1)) * math.cos(toRad(lat2)) * math.sin(dLon/2) * math.sin(dLon/2)
  c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
  d = R * c
  return round((d + 2.220446049250313e-16) * 100) / 100

if __name__ == "__main__":
  argc = len(sys.argv)
  if argc<5:
    print(f"Default coords. argc = {argc}")
    lat0 = 22.4
    lat1 = 22.4
    long0 = 113.9
    long1 = 114.0
  else:
    print(f"Custom coords. argc = {argc}")
    lat0 = float(sys.argv[1])
    long0 = float(sys.argv[2])
    lat1 = float(sys.argv[3])
    long1 = float(sys.argv[4])
  km = haversine(lat0, long0, lat1, long1)
  mi = haversine(lat0, long0, lat1, long1, False)
  print(f"Distance between {lat0}, {long0} and {lat1}, {long1}:")
  print(f" * {km} km")
  print(f" * {mi} mi")
